### PR TITLE
Add picker options: filter, select_ops

### LINKED ISSUE #14: Add condition function to filter commands
# Backround
Some commands aren't applicable for some file types or some Vim modes. For example:
- Sort JSON command for Lua buffer
- Copy selection for the normal mode
- Run git command for not versioned directory

# Proposal
Make the commands list more flexible. Decide which command should be displayed before the picker opening.
Pseudocode:
```lua
opts = {
  commands = {
    { name = "Copy selection", execute = ":y", tags = { "Editor"}, modes = { "v" } },
    { name = "Sort JSON", execute = ":%!jq", tags = { "Editor" }, filetypes = { "json" } },
  },
  filter = function(command)
   -- If command filetypes contains json and buffer isn't json, then return false
   -- If the current mode is normal, then "Copy selection" should be hidden. Thus, the function returns false
   return true
  end
}
```

The command list should be re-filtered on the picker opening.

# Note
This functionality can replace the property `tags`
```lua
local current_tags = nil

-- Configuration
opts = {
  filter = function(command)
    -- function takes current_tags into account
  end
}

-- Usage
current_tags = { "git" }
require("toolbox").open_picker()
current_tags = { "json" }
require("toolbox").open_picker()
current_tags = nil -- or {}
require("toolbox").open_picker()
```

### PR DESCRIPTION
**Pull request by fpohtmeh**
This PR introduces more flexibility to the command picker
- Specifying the filter function for the commands list
- Specifying `vim.ui.select` options
- Adding **Recipes** section to explain how to configure the toolbox for different needs

As a result, the property `tags` was removed because the new mechanism covers it. Please check the example with tags in **Recipes** section

This PR closes #14
Also, with the `vim.ui.select` options, it's easy to format the command name and append the description (#11)

### CODE REVIEW

**Comment by DanWlker** [comment]
Hi thank you for the PR.

For the tags, I am still having trouble understanding how this covers other scenarios. For the filetype usages that you showed, you could just check the filetype when running the keymap and invoke the correct tag right?

Take note that tags is a list, therefore one command can be associated with multiple tags.

As an example

```lua
 {
      '<leader>st',
      function()
        require('toolbox').show_picker(vim.bo.filetype)
      end,
      desc = '[S]earch [T]oolbox',
      mode = { 'n', 'v' },
    },
```

For the format_item that looks like an interesting idea, but can we split it into an issue and have a discussion there first? Thanks

**Comment by fpohtmeh** [comment]
@DanWlker 1st parameter of the `show_picker` is always a function for filtering. Thus, your example is not correct. This is the corrected version
```lua
 {
      '<leader>st',
      function()
        require('toolbox').show_picker(function(command) command.filetype == vim.bo.filetype end)
      end,
      desc = '[S]earch [T]oolbox',
      mode = { 'n', 'v' },
},
```
The function for tags is different, and it's available in recipes.
Please provide me a scenario which isn't clear for you, and I'll be glad to explain

---

I'm OK with moving `format_item` to another PR. On the other hand, it's a very small scope of changes, and it's documented in builtin `vim.ui.select`, so I decided to include in this PR

**Comment by DanWlker** [comment]
Err no I'm not referring to your changes, I'm referring to the current implementation. 

My question is: Is there a scenario where the current implementation cannot fulfill. The example I gave above shows how you would filter by file type using tags

**Comment by fpohtmeh** [comment]
The current implementation is designed for any set of custom properties in the command, and I don't see limitations.

I prefer how the user adds custom-named properties to commands and uses them in the filtering.
Some predefined properties, like `tags` or `filetype,` may push the user to set these properties even if they don't apply to the user's usage scenario.

**Comment by DanWlker** [comment]
I will close this in an hour's time since I don't see a scenario that the current implementation on `main` cannot fulfill. We can explore this again if we hit a wall in the future, but for now I don't want to do an at most "side-grade" implementation. It is also why the ideas in the issues are all just in consideration as most of them are either a "side grade" or do not have tangible use cases that the current implementation cannot fulfill

Thanks again for spending the time on the PR. I am happy to further explore the vim ui select implementation if you would like to go for that

**Comment by fpohtmeh** [comment]
Hm, AFAIU you would like to use tags as universal properties for the filtering.
However, it has many limitations, and it's not always obvious how to use tags to cover some use cases.

Use case which can't be covered with tags in the future:
- Display 10 last used commands

**Comment by DanWlker** [comment]
Yup, that sounds valid. I'll do a code review later.

**Comment by DanWlker** [comment]
Side note, I'm thinking of keeping tags around as a 'pre-filter'/grouping as I'm not sure how fast is the filtering when it gets to many items and it is done on the fly

(Though it is reasonable to also assume the amount of custom commands that a user has should not be alot)

Do you have any inputs

**Comment by fpohtmeh** [comment]
The filtering is quick, O(n) complexity

File: lua/toolbox/init.lua [hunk 1]
```diff
@@ -45,39 +62,29 @@ function M.run(name)
...
+	select_opts = vim.tbl_extend("force", {
+		prompt = "Toolbox",
+		format_item = function(command)
+			return command.name
+		end,
+	}, select_opts or {})
 
-		local execute = M.commandMap[choice].execute
-		if execute == nil then
+	vim.ui.select(H.filter_commands(filter), select_opts, function(command)
```
Comment (DanWlker) [inline lua/toolbox/init.lua:80]: Move this to a different PR, I like this but lets land the filtering first

File: README.md [hunk 2]
```diff
@@ -113,6 +105,73 @@ return {
...
+```
+
+### Usage
+
+```lua
+-- Show only clipboard commands
+local function filter(command)
+  return command.tags ~= nil and vim.tbl_contains(command.tags, "clipboard")
+end
+require("toolbox").show_picker(filter, { prompt = "Select clipboard command" })
```
Comment (DanWlker) [inline README.md:131]: The prompt is also a good idea, just not this PR

File: README.md [hunk 3]
```diff
@@ -113,6 +105,73 @@ return {
...
+require("toolbox").show_picker(nil, {
+  format_item = function(command)
+    -- Display => and execute string after the name
+    return command.name .. " => " .. (type(command.execute) == "function" and "<function>" or command.execute)
+  end
+})
+```
+
+</details>
+
```
Comment (DanWlker) [inline README.md]: Add one example for the filtering by 10 last used commands

File: lua/toolbox/init.lua [hunk 4]
```diff
@@ -12,29 +38,20 @@ function M.setup(opts)
...
-	local execute = M.commandMap[name].execute
+	local command = H.find_command(name)
+	if command == nil then
+		error("Command " .. name .. " is not found", 0)
+	end
+
+	local execute = command.execute
 	if execute == nil or type(execute) ~= "function" then
-		error("Unknown or unexecutable command", 0)
+		error("Command has unsupported function type", 0)
```
Comment (DanWlker) [inline lua/toolbox/init.lua]: Revert this

File: lua/toolbox/init.lua [hunk 5]
```diff
@@ -12,29 +38,20 @@ function M.setup(opts)
...
-			end
-			table.insert(M.tagToCommandList[tag], command.name)
-		end
-	end
+	H.opts = opts
 end
 
 function M.run(name)
-	local execute = M.commandMap[name].execute
+	local command = H.find_command(name)
```
Comment (DanWlker) [inline lua/toolbox/init.lua:45]: Is there a reason why we are changing it to a list?

File: lua/toolbox/init.lua [hunk 6]
```diff
@@ -45,39 +62,29 @@ function M.run(name)
 	}
 end
 
-function M.show_picker(tag)
+function M.show_picker(filter, select_opts)
```
Comment (DanWlker) [inline lua/toolbox/init.lua:65]: Move select_opts to a different PR

File: lua/toolbox/init.lua [hunk 7]
```diff
@@ -45,39 +62,29 @@ function M.run(name)
...
-		promptTitle = promptTitle .. " <" .. tag .. ">"
-	end
-
-	vim.ui.select(M.tagToCommandList[tag], {
-		prompt = promptTitle,
-	}, function(choice)
-		if choice == nil then
-			return
-		end
+	select_opts = vim.tbl_extend("force", {
```
Comment (DanWlker) [inline lua/toolbox/init.lua:73]: This as well

File: lua/toolbox/init.lua [hunk 8]
```diff
@@ -105,23 +112,22 @@ function M.show_picker(tag)
...
+			if command.require_input then
 				vim.api.nvim_feedkeys(cmdPrefix .. execute, "m", false)
 				return
 			end
 			local ok, res = pcall(vim.cmd, cmdPrefix .. execute)
 			if not ok then
 				error(res, 0)
 			end
+		else
+			error("Command has unsupported function type", 0)
```
Comment (DanWlker) [inline lua/toolbox/init.lua]: Change to "Unknown or unexecutable command"

**Comment by DanWlker** [comment]
It does make sense to land the select opts as well, since it is used to fill in the name of the command (I learned something new about vim.ui.select as well today.) Just add the example for filtering by 10 last used commands and we can land this, thanks!

I tested the filtering, works quick with a high number of items so LGTM

**Comment by DanWlker** [comment]
> Side note, I'm thinking of keeping tags around as a 'pre-filter'/grouping as I'm not sure how fast is the filtering when it gets to many items and it is done on the fly

I tested it on my device, it is still instant even with 5k items, will deprecate this approach

File: README.md (hunk 3 above)
Comment (fpohtmeh) [reply README.md]: 10 last commands feature needs extra functionality.
First of all, I need to introduce callback:
```lua
opts = {
  commands = {},
  pre_run = function(command)
    -- save timestamp for the executed command to some table
  end,
  post_run = function(command, success)
    -- added for consistency with pre_run
    -- potential use case - notify about the executed command name, success or failure (info or warning)
  end
}
```

After this, I should save usage timestamps in the data folder, to be available in the next session.

Let's do this in another PR

Comment (DanWlker) [reply README.md]: hmm, this is a good point, are there any other scenarios that do not require this and that the 'tags' implementation in main cannot support?

Comment (fpohtmeh) [reply README.md]: The problem with tags is that the user should manually set them, and tags are a limited set of values. With tags, the user can't filter by integers, double values, or dynamically calculated values (like the last usage timestamp).

More examples that can't be covered with tags:

1. Display commands that were never used
2. Display the ten most popular commands
3. Display commands that were applied to the current project (or the current directory) earlier
And more ...

Note that examples 1-3 are similar to the "10 recent" case and need `pre_run` callback and storing session data

Comment (DanWlker) [reply README.md]: Let me think if I want to land this without pre_run and post_run, since all the listed scenarios that the change covers require that.

Also they probably should be attached to `show_picker` instead of opts, however the presence of `require_input` also complicates thing abit

Comment (fpohtmeh) [reply README.md]: Examples of filters without the `pre_run`:
- Show commands that start with "Copy" in the name.
- Show commands with Lua functions only.
- Show commands with the word "Format".
- Show commands with 1 or more arguments.

In all cases, it's easier to write a universal function than set tags and maintain them.

Comment (DanWlker) [reply README.md:174]: If you'd like, we can land the vim ui select first in a different PR. I am still on the fence a bit for this

Comment (fpohtmeh) [reply README.md]: I'm OK to merge select_opts first

Comment (DanWlker) [reply README.md]: Lets merge select_opts first. I am very thankful for you spending your time on this but give me a few days to think through this.

Comment (fpohtmeh) [reply README.md]: Please remember that if the plugin is oriented on custom properties, it can be used in filtering and sorting. Tags are unusable for sorting

Comment (DanWlker) [reply README.md]: I will think about it. Thank you.

Comment (DanWlker) [reply README.md]: Hi @fpohtmeh, I've given it a thought. I would like to request to do this enhancement in a new function, perhaps `show_picker_custom()` or `show_picker_advanced()` so that for advanced users that want advanced filtering and or sorting in the future, they can use that, while we remain the current behaviour for the default show_picker.

I can help implement it, the default `show_picker()` will just call show_custom_picker under the hood.

What do you think of this idea?

Comment (DanWlker) [reply README.md]: image

Comment (DanWlker) [reply README.md]: If you are ok with this, I will push the changes, update the documentation and proceed to merge it.

Comment (fpohtmeh) [reply README.md]: Yes, please proceed

Comment (fpohtmeh) [reply README.md]: But my opinion about the tag is the same - it covers some of your specific use cases.
Users lose nothing if you remove tags but introduce a filter and a recipe with a tag. And the plugin API will be simpler

**Comment by DanWlker** [comment]
@fpohtmeh I made some changes, can you help to see if you have any further comments? if no I'll proceed to merge. Thank you

File: README.md [hunk 9]
```diff
@@ -113,6 +113,64 @@ return {
...
+    { name = "Format JSON with jq", execute = ":%!jq", filetype = "json" }
+    { name = "Format QML file", execute = ":qmlformat %", filetype = "qml" }
+  },
+}
+```
+
+#### Usage
+
+```lua
+require("toolbox").show_picker({
```
Comment (fpohtmeh) [inline README.md]: Use the show_picker_custom. Please fix the other recipes

File: lua/toolbox/init.lua [hunk 10]
```diff
@@ -45,85 +72,114 @@ function M.run(name)
...
+
+	M.show_picker_custom({
+		filter = function(command)
+			return command.tags ~= nil and vim.tbl_contains(command.tags, tag)
+		end,
+	}, select_opts)
+end
+
+---@param opts Toolbox.ShowPickerCustomOpts?
+---@param select_opts table Taken from vim.ui.select
```
Comment (fpohtmeh) [inline lua/toolbox/init.lua:101]: You may reffer to vim.ui.select opts in annotation

File: lua/toolbox/init.lua [hunk 11]
```diff
@@ -45,85 +72,114 @@ function M.run(name)
...
-		end
 
-		if type(execute) == "string" then
-			local cmdPrefix = ":"
-			if isVisual then
-				cmdPrefix = cmdPrefix .. "'<,'>"
-			end
+			local execute = command.execute
+			if type(execute) == "function" then
+				local numParams = debug.getinfo(execute).nparams
```
Comment (fpohtmeh) [inline lua/toolbox/init.lua]: it's better to use underscore in Lua names, not camelcase

Comment (fpohtmeh) [reply lua/toolbox/init.lua]: Check other variables for this errror

File: lua/toolbox/init.lua (hunk 10 above)
Comment (DanWlker) [reply lua/toolbox/init.lua:101]: yeah this is actually copied from there

image

**Comment by DanWlker** [comment]
Last check. I will merge this some time tomorrow if all is ok. Thank you again for the PR

**Comment by fpohtmeh** [comment]
looks good
